### ISSUE TITLE
Implausible cycle count when PMU muxing is active

### ISSUE THREAD

**Issue by topolarity** [JuliaPerf/LinuxPerf.jl#53]
It's not uncommon for me to get a reading like this:
```julia
julia> @pstats "cpu-cycles,stalled-cycles-frontend,stalled-cycles-backend,instructions,branch-instructions,branch-misses" rand(1000,1000)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
╶ cpu-cycles               1.03e+05   46.6%  #  0.0 cycles per ns
╶ stalled-cycles-frontend  4.16e+03  100.0%  #  4.0% of cycles
╶ stalled-cycles-backend   3.58e+04  100.0%  # 34.6% of cycles
╶ instructions             1.28e+07  100.0%  # 124.1 insns per cycle
╶ branch-instructions      1.12e+06  100.0%  #  8.7% of insns
╶ branch-misses            2.22e+04   53.7%  #  2.0% of branch insns
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```

about ~10-30% of the time, compared to a more reasonable measurement:
```julia
julia> @pstats "cpu-cycles,stalled-cycles-frontend,stalled-cycles-backend,instructions,branch-instructions,branch-misses" rand(1000,1000)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
╶ cpu-cycles               9.21e+06   33.3%  #  1.5 cycles per ns
╶ stalled-cycles-frontend  8.19e+03   74.9%  #  0.1% of cycles
╶ stalled-cycles-backend   4.69e+04  100.0%  #  0.5% of cycles
╶ instructions             1.28e+07  100.0%  #  1.4 insns per cycle
╶ branch-instructions      1.13e+06  100.0%  #  8.8% of insns
╶ branch-misses            1.49e+04   92.0%  #  1.3% of branch insns
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```

124 instructions per cycle seems... a little high

**Comment by topolarity** [comment]
Note this is even with the diff from https://github.com/JuliaPerf/LinuxPerf.jl/issues/50#issuecomment-2554247942 applied, so it could be that something else is up here

**Comment by topolarity** [comment]
I haven't been able to get `perf stat` to make the same mistake so LinuxPerf may be doing something wrong here:
```
$ perf stat -e cycles,instructions,branches,branch-misses,cache-references,cache-misses echo


 Performance counter stats for 'echo':

         1,417,515      cycles                                                        (81.44%)
         1,048,902      instructions              #    0.74  insn per cycle
           240,734      branches
            31,285      branch-misses             #   13.00% of all branches
            87,314      cache-references
            25,054      cache-misses              #   28.694 % of all cache refs      (18.56%)

       0.001116123 seconds time elapsed

       0.000000000 seconds user
       0.001277000 seconds sys
       ```

**Comment by topolarity** [comment]
Seems to be affected by the number of events we're creating for threading (I'm running on a 64-core / 128-hyper-threads machine). This is the worst outlier with `threads=false` which is orders of magnitude improved:
```julia
julia> @pstats "cpu-cycles,instructions,branch-instructions,branch-misses,cache-misses,cache-references" threads=false rand(1000,1000)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
╶ cpu-cycles               4.27e+06   73.0%  #  1.2 cycles per ns
╶ instructions             1.03e+07  100.0%  #  2.4 insns per cycle
╶ branch-instructions      6.23e+05  100.0%  #  6.0% of insns
╶ branch-misses            4.94e+03  100.0%  #  0.8% of branch insns
╶ cache-misses             8.85e+04  100.0%  #  7.6% of cache refs
╶ cache-references         1.17e+06   27.4%
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```

Using `OPENBLAS_NUM_THREADS=1` also dramatically reduces the noise, since most of the threads we're registering perf counters for are OpenBLAS's

Weirdly enough, the situation actually _improves_ if I use `-t auto` (with `threads=true`)

**Comment by topolarity** [comment]
If I group the `cpu-cycles` together with the `instructions` then both estimates seem to have these very low readings:
```julia
julia> @pstats "(cpu-cycles,instructions),branch-instructions,branch-misses,cache-misses,cache-references" rand(1000,1000)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
┌ cpu-cycles               8.71e+04   40.7%  #  0.0 cycles per ns
└ instructions             2.60e+04   40.7%  #  0.3 insns per cycle
╶ branch-instructions      1.01e+06  100.0%  # 3900.4% of insns
╶ branch-misses            1.33e+04  100.0%  #  1.3% of branch insns
╶ cache-misses             1.04e+05  100.0%  # 11.7% of cache refs
╶ cache-references         8.91e+05   59.6%
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```

Somehow these counters were active ~40% of the time, but the reading was ~100x lower than expected

**Comment by Zentrik** [comment]
All the issues you've had have been on a single machine right?

**Comment by topolarity** [comment]
Yeah, that's right. I'm going to try to take some measurements tomorrow on other HW.

**Comment by topolarity** [comment]
OK, it looks like there may be two sources of noise here...

Above I had P-states enabled and used `enable_all!` / `disable_all!`.

On the other hand, if I enable normal P-states and use `enable!` / `disable!` (instead of `disable_all!`):
```julia
julia> @pstats "cpu-cycles,instructions,branch-instructions,branch-misses,cache-misses,cache-references,L1-dcache-loads,L1-dcache-load-misses,L1-dcache-stores,L1-dcache-loads" rand(1000,1000)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
╶ cpu-cycles               3.73e+04   47.2%  #  0.0 cycles per ns
...
```

So we get bad readings for both:
  - `enable!` / `disable!` with lower-energy P-states enabled
  - `enable_all!` / `disable_all!` with processor locked to `P-0` / `C-0`

The effects of having P-states enabled on `cpu_cycles` is generally much bigger (e.g. above measurement), but using `disable_all!` seems to cause bad / zero readings (less often, but pretty reliably).

Both of these only seem to happen with `threads=true` (the default). I've tried to create a C-reproducer that shows the same problem, but so far I can't get it to show a problem nearly as severe.

**Comment by topolarity** [comment]
OK I think I have a root cause.

As it turns out, our `kernel` overhead for enabling / disabling perf events is _way_ bigger than I thought. Across the several ~hundred perf `fd`'s that we open (primarily for the ~32 openBLAS threads on my machine), it takes 1M instructions / 10M cycles / 7 milliseconds to enable and disable the perf events once. If `threads=false`, then that overhead drops to 20k instructions / ~0.2 milliseconds, so it seems to scale linearly with the number of `fd`s we have open, and that overhead is pretty much the same regardless of whether we use `disable!` or `disable_all!`.

So the time we shaved off in https://github.com/JuliaPerf/LinuxPerf.jl/pull/38 is all on the user side, but the time in-kernel overhead _strongly_ dominates. That would generally be OK, because perf is supposed to ignore time spent in kernel (if you set `exclude_kernel` as we do). However, as it turns out this is not true for `time_running` / `time_enabled` (see https://lore.kernel.org/linux-man/CAAM_cidrPK1W+K-nb1gY_QHbveKguOMzG34NJ=_QKhz49=vh9A@mail.gmail.com/). Those estimates always include time spent in the kernel! So the basic problem here is that we are dominated by time spent in the kernel, which skews our running time estimates and making our scaled counts _much_ less reliable when `exclude_kernel` is being used.

So that's the explanation. I'm not sure what that means we should do to adapt / workaround it yet... any thoughts @vchuravy or @Zentrik ?

**Comment by topolarity** [comment]
> Do other tools like perf deal with this in anyway?

Not to my knowledge - when used externally you need a relatively short run (~few milliseconds) with significant kernel overhead to notice this, so it'd be pretty rare

It's unfortunately quite common for in-process monitoring, since enabling / disabling the events guarantees a significant amount of kernel overhead and you're more likely to be measuring a short ~few ms event.

**Comment by topolarity** [comment]
Perhaps we should run with `threads=false` by default?
